### PR TITLE
The error.* axis: an invalid-method coverage surface for the MTHDS Test Corpus

### PR DESCRIPTION
**Pull request by lchoquel**
The MTHDS Test Corpus covered the valid-method space — native concepts, operators, controllers, hand-named features. This adds the invalid half: a generated `error.*` tag namespace, focused entries covering it, and the gate that keeps the two honest. It implements Phase 3.3 of the corpus tracker, from the precondition audit that preceded it.

## The `error.*` namespace

The namespace generates from `VALIDATION_ERROR_TYPES`, the registry landed by #1138, and needs **no normalization rule of its own**. That registry carries two spellings — stage codes that are already snake_case (`missing_input_variable`) and the dry-run residual as an exception class name (`DryRunError`) — and the corpus's existing three-step normalization absorbs both: it leaves the first untouched and turns the second into `dry_run_error` exactly as it turns `YesNo` into `yes_no`. `code` records each registry spelling verbatim, as it already does everywhere else.

**Five codes are excluded, and each exclusion was measured rather than argued.** The rule applied throughout: write the bundle that ought to trigger the code, run it, and record what the runtime actually said. Reading the call graph would have been wrong in both directions.

- `circular_dependency_error` — raised only from the pipe sorter, which runs only on `PipelexBundleSpec.to_blueprint()`. A `.mthds` file parses straight into a blueprint and never passes through it.
- `unknown_concept` — shadowed. The pipe factory's output check fires under exactly the condition the earlier concept-reference resolution already rejects, and reports the same fault. Measured across unqualified, cross-domain and qualified-but-missing output refs: every one comes back as `unresolved_concept`.
- `optional_force_redundant` — advisory-only; it rides the `warnings` array and never makes a verdict invalid. The manifest has no warnings axis, so there is no shape a focused entry could take.
- `unknown_validation_error` and `unknown_factory_error` — the generic fallbacks. An entry pinning one would be pinning the absence of a diagnosis.

## The entries

Twenty focused invalid entries, each verified to produce exactly one error of exactly its declared type. Eighteen are absorbed from `conformance/validate-error-qa/cases/`; two (`invalid_missing_pipe_type`, `invalid_native_concept_redeclaration`) are new, written to close registry codes no existing case exercised.

Every absorbed bundle is re-themed on the way in, per the corpus's subject-matter rule: crate labels, parcel routing, ferry crossings, allotment plots. The entry *names* keep their machinery words — that is the rule's carve-out, since an invalid entry is named for the defect it triggers — but the descriptions and the bundles underneath are ordinary-world. Model pins are dropped: an entry that names a model is an entry about model selection.

## The agreement gate

An invalid entry states its fault twice in one file: as `expected_error` (the wire string) and as an `error.*` tag in `covers` (the normalized form). Nothing checked that the two agree, so an entry could claim coverage of a tag it does not produce — and the exhaustivity gate would count the claim. The third arm requires the agreement, joining through the vocabulary's own `code` field so the gate carries no second copy of the normalization rule.

## Two fixes the work uncovered

**`fix(validation)`: structured error data was destroyed when a bundle loaded from a library directory.** Running the conformance cases through the local runtime returned untyped residuals where the committed API artifacts carry real `error_type` values. Not API-vs-local skew — two re-raise arms in `LibraryManager.load_library_from_mthds_paths` were lossy. Wrapping a `ValidationError` in a bare `LibraryError` routed it to the arm that *forwards* structured data, which had none to forward; and the reference-error arm rebuilt the exception without carrying its payload across. So the same bundle produced different diagnostics depending on which of `validate_bundle`'s two entry shapes it arrived through: the in-memory `mthds_contents` path (what the HTTP surface uses) was typed, the `library_dirs` path (what the CLI and the corpus gate use) was not. `unresolved_concept`, `optional_input_unguarded` and `optional_output_required` were simply unreachable from a file on disk.

This was invisible because every existing structured-error test used the contents shape. The new parity test validates one bundle through **both** shapes and compares each against an explicit expected list, so "both degrade together" cannot satisfy it. It fails on two of three cases without the fix.

**`fix(tests)`: the `validate --all` smoke test had been sweeping the corpus without saying so.** It scanned the `pipelex/` package directory, chosen to avoid conflicting fixtures under `tests/` — but the package's only `.mthds` files are the corpus's. The first invalid entry turned it red. It now selects the corpus's `valid` entries explicitly, which states the intent the path expression only implied.

## Two runtime defects found and deliberately not fixed

Both surfaced from the reachability probes, and both are recorded in the tracker rather than fixed here.

**A mutually recursive `PipeCondition` pair crashes the validator.** Two conditions each routing to the other exhaust the interpreter's stack during the dry-run sweep — `RecursionError`, traceback dominated by `pipelex/pipe_controllers/pipe_condition.py`. The user-visible result is a crash rather than a verdict, which is the one thing the validate surface is supposed never to do. A self-referencing `PipeSequence`, by contrast, fails cleanly as `DryRunError`.

**`circular_dependency_error` is unreachable from any `.mthds` file** — the same underlying gap, and the reason for its exclusion above.

Both want cycle detection on the parse path, which is a real feature with its own design, tests and spec impact. Filed rather than smuggled in here.

## One config change

`invalid_missing_pipe_type` and `invalid_unknown_pipe_type` fail `plxt lint`'s schema check — which is exactly what they exist to do. They are added to the **global** `exclude` list in `.pipelex/plxt.toml` (a per-rule `exclude` does not lift the schema check; measured). The corpus's own entry-validation gate is strictly stronger over these files: it requires each to fail with exactly its declared error, where lint only requires it to fail.

## Gates

`make agent-check` and the full `make agent-test` both green on this tree.

## Contract

The workspace spec `docs/specs/mthds-test-corpus.md` is updated in the workspace repo, on `feature/Corpus-4`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Adds the corpus’s invalid-method axis via a generated `error.*` tag namespace and 20 focused entries, and fixes validation parity so diagnostics match whether bundles come from in-memory contents or library directories. Tightens the exhaustivity gate: an invalid entry’s `covers` must exactly match its `expected_error`, and valid entries can no longer claim any `error.*` tag.

- Vocabulary: generator now emits `error.*` from `VALIDATION_ERROR_TYPES` (existing normalization; `DryRunError` → `error.dry_run_error`), records five measured exclusions, and updates `vocabulary.toml`.
- Entries: 20 focused invalid entries (18 migrated, 2 new: missing pipe type, native concept redeclaration), each producing exactly one declared error; `covers` contains only the matching `error.*` tag.
- Gate behavior change: the exhaustivity check now requires exact agreement — `covers` must be exactly [the `error.*` tag] whose `code` equals `expected_error`; previously membership allowed extra tags. A symmetric guard now rejects `error.*` tags on valid entries.
- Runtime fix: library-directory validation now categorizes `ValidationError` before wrapping and preserves `ConceptLibraryError` payloads; a parity test pins that contents-based and library-dir validations report the same categorized items. No wire values changed. Comments and docs corrected to `LibraryManager._load_mthds_files_into_library`.
- CLI/tests/lint: `validate --all` assembles and validates only the corpus’s valid entries and asserts a non-empty set; invalid bundles are globally excluded from schema lint in `.pipelex/plxt.toml`; contributor docs and changelog updated.

Required migration for corpus authors:
- For invalid entries, set `expected_error` to a registry value and make `covers` exactly the corresponding `error.*` tag — remove any non-error tags.
- Do not add `error.*` tags to valid entries.

<sup>Written for commit 43f49fd64f76b61d9afd8c0699a4386e9c4a85e5. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Pipelex/pipelex/pull/1139?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: tests/unit/pipelex/test_extras/test_mthds_corpus_exhaustivity.py [hunk 1]
```diff
@@ -29,3 +36,29 @@ def test_every_tag_an_entry_covers_exists_in_the_vocabulary(self) -> None:
...
+        }
+        disagreements: list[str] = []
+        for entry in iter_entries(validity=EntryValidity.INVALID):
+            expected_error = entry.manifest.expected_error
+            assert expected_error, f"Corpus entry '{entry.name}' is invalid, so the manifest guarantees an expected_error"
+            owed_tag = tag_by_error_code.get(expected_error)
+            if owed_tag is None:
+                disagreements.append(f"{entry.name}: expected_error '{expected_error}' is in no error.* tag's `code`")
+            elif owed_tag not in entry.manifest.covers:
+                disagreements.append(f"{entry.name}: expected_error '{expected_error}' means it must cover '{owed_tag}'")
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply tests/unit/pipelex/test_extras/test_mthds_corpus_exhaustivity.py]: ✅ **Fixed** — confirmed, and the gap was real. `test_every_required_tag_has_a_focused_entry` calls `iter_entries(granularity=EntryGranularity.FOCUSED)` with no `validity` argument, and `loader.py:121` applies the validity filter only when one is passed — so invalid entries do count toward coverage. That is necessary rather than accidental: the 19 required `error.*` tags are covered by nothing but invalid entries. It also means membership was not enough, exactly as you say.

The gate now compares `entry.manifest.covers != [owed_tag]`, and the assertion message says why an invalid entry's `covers` is its error tag and nothing else. Equality makes the hole unreachable — an invalid entry can no longer contribute a `native.*` / `operator.*` / `controller.*` / `feature.*` tag at all — so the coverage computation itself needed no change.

Mutation-tested rather than argued. Moving `"operator.llm"` from `entries/native_text_delay_notice/entry.toml` onto `entries/invalid_missing_pipe_type/entry.toml` left all three arms green with `PipeLLM` no longer covered by any working fixture; with the equality change that run fails, naming the entry and stating what to do; reverting the fixtures returns it to green.

Two alternatives were considered and dropped. Filtering coverage to VALID entries reds all 19 error tags immediately. A manifest-model validator has no vocabulary access, so it could only assert `len(covers) == 1` plus an `error.` prefix, splitting the rule across two places and raising a bare pydantic error instead of the gate's explanatory one.

`docs/contribute/mthds-test-corpus.md` was sharpened from "does not contain" to "is not exactly" in the gate table; the normative sentence already said equality. The matching line in the workspace spec is filed as a cross-repo follow-up.

File: .pipelex/plxt.toml [hunk 2]
```diff
@@ -24,6 +24,14 @@ exclude = [
...
   ".git/**",
   "*.lock",
+  # The MTHDS Test Corpus's deliberately invalid entries — the one .mthds population this repo does
+  # not lint, because each one exists to be rejected. Their fault is sometimes structural (a typeless
+  # pipe section, an unknown pipe `type`), which the schema catches as the error the entry was written
+  # to trigger. What guards them instead is strictly stronger: the corpus's entry-validation gate runs
+  # the real validation engine over every entry in CI and is red both when an invalid entry fails
+  # differently and when it accidentally validates. Excluded rather than schema-exempted because a
+  # per-rule `exclude` does not lift the schema, only narrows which rule applies.
+  "**/mthds_corpus/entries/invalid_*/*.mthds",
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply .pipelex/plxt.toml:34]: ➖ **False positive** — the count is right, the conclusion is not. Each half was measured rather than reasoned.

**The count checks out.** Running `plxt lint --no-auto-config --schema-path derived/mthds_schema.json` over all 20 invalid bundles, exactly two trip the schema: `invalid_missing_pipe_type` (`"type" is a required property`) and `invalid_unknown_pipe_type` (`"PipeWondrous" is not one of [...]`). The other 18 lint clean.

**No formatting coverage is lost.** `plxt format --check` over all 20 reports them already byte-exact under the `.mthds` rule's options. The blocks that look unaligned are canonical output — taplo skips alignment for runs whose value pushes past `column_width = 80`. So `merge-check-plxt-format` was doing no work on this population and the exclude removes nothing.

**The remaining lint coverage is subsumed by a strictly stronger gate.** What is left is TOML syntax, duplicate keys, and `additionalProperties`. Each of those makes the bundle produce a *different* diagnostic — and `tests/integration/pipelex/test_extras/test_mthds_corpus_entries.py:51` asserts exact list equality against `expected_error`, so a widened blast radius or a changed fault is red in CI. Only the precision of the failure message is coarser, never the guarantee.

**The narrower alternative does not exist in this tool.** A per-rule `exclude` does not lift the schema — `SchemaAssociations::new` unconditionally calls `add_builtins`, which registers `.*\.mthds$` against the binary's bundled schema at `priority::BUILTIN`, so dropping the config rule just demotes to a staler schema that still rejects a typeless pipe. And `[rule.schema] enabled = false` is a project-wide kill switch rather than a scoped one (that turned out to be a genuine plxt bug, now filed against `vscode-pipelex`). The global `exclude` is the only lever; the only question is how wide.

**Two entries would be a less stable list than the wildcard.** `derived/mthds_schema.json` is regenerated from this repo's own blueprints on every `make plxt-lint`, so which entries trip the schema is a moving property of our code, not a fixed fact. The moment someone constrains a pipe-code or concept-ref string — plausible hardening — `invalid_pipe_code_syntax`, `invalid_optional_marker_invalid` or `invalid_native_concept_redeclaration` start reddening `make lint` on files nobody touched, reporting that a deliberately-invalid entry is invalid. The wildcard states the policy; a two-item list encodes today's accident.

One genuine residue, and thank you for it: the contributor docs warned that the editor hook fights invalid entries but never said those files are also outside `make lint` and `make format`. `docs/contribute/mthds-test-corpus.md` now says so.

File: tests/integration/pipelex/pipeline/test_validate_bundle_entry_shape_parity.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+from pathlib import Path
+
+import pytest
+
+from pipelex.base_exceptions import ValidationErrorCategory
+from pipelex.pipeline.exceptions import ValidateBundleError
+from pipelex.pipeline.validate_bundle import validate_bundle
+from pipelex.validation_error_types import PipeValidationErrorType
+
+_UNRESOLVED_CONCEPT_MTHDS = """
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply tests/integration/pipelex/pipeline/test_validate_bundle_entry_shape_parity.py:32]: ➖ **False positive** — the rule is quoted accurately, but the tree does not follow it for this kind of data, and following it here would make this file the outlier in its own directory.

**Measured on this branch:** 90 test modules inline an MTHDS bundle literal. **Zero** of them keep it in a `test_data.py`. Not "few" — none. In this file's own directory, 32 of the 33 modules inline one. Of the 52 `test_data.py` files in the tree, the only one holding anything MTHDS-shaped is `tests/unit/pipelex/graph/test_data.py`, and it holds GraphSpec dicts imported by two sibling modules.

That last point is what reveals the operative norm, and it is a coherent one: **`test_data.py` is where data shared across modules goes.** The three payloads here are `_`-prefixed private constants, each consumed exactly once, in the `pytest.param` list about sixty lines below its own definition, and referenced nowhere else in `tests/` or `pipelex/`. Moving them would put a wall between each bundle and the `(category, error_type)` tuple that is its entire point.

The nearest sibling settles it: `tests/integration/pipelex/pipeline/test_validate_bundle_structured_errors.py` pre-exists on `dev` and is structurally identical — module docstring, then `_UNKNOWN_CONCEPT_MTHDS` / `_MISSING_DEPENDENCY_MTHDS` inline constants, then one parametrized test class. This new file is a deliberate mirror of the module it extends.

The rule as written is worth revisiting rather than applied here, since it has now been raised on three consecutive PRs. It traces to `pipelex/kit/agent_rules/pytest_standards.md:17`, which we ship to users — so it is a dogfooding question, not a per-PR one. Captured with the measurements in `wip/pr-1139-review-notes.md`, alongside the two prior rounds.

File: tests/integration/pipelex/pipeline/test_validate_bundle_entry_shape_parity.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
+"""Pin: the same bundle produces the same diagnostics whichever way it enters ``validate_bundle``.
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply tests/integration/pipelex/pipeline/test_validate_bundle_entry_shape_parity.py:1]: ➖ **False positive** — on two grounds, one about the population and one about what the rule targets.

**The population.** Counting with `ast.get_docstring` on the module node: **691 of 1130 test modules (61%)** open with a module-level docstring. By directory — 32 of 33 in `tests/integration/pipelex/pipeline/`, 9 of 13 in `tests/e2e/pipelex/cli/`, 4 of 4 in `tests/unit/pipelex/test_extras/`. The other half of the same sentence fares the same: 553 of 1216 `Test*` classes carry a class docstring. Nothing enforces any of it — `pyproject.toml` selects `ALL` but ignores `D100` and `D101`, and no rule forbids a module docstring. The pre-existing sibling `test_validate_bundle_structured_errors.py` on `dev` opens with a longer module docstring of precisely this kind.

**What the rule targets.** It forbids "this test tests X" boilerplate migrating up to file level, and this file already satisfies its positive half — the concise test-purpose docstring is on the method (`"""Both shapes report the expected categorized items — and therefore each other's."""`). The module docstring is doing different work: it records which two entry shapes exist, that `load_library_from_mthds_paths` destroyed structured error data, and why a third already-parity bundle sits in the parametrize list, so that a fix degrading *both* shapes still fails. That last sentence is the most valuable thing in the file and has no natural home on a single parametrized method.

Conforming this file would satisfy the letter of the rule while making it the only module in its directory that behaves differently.

Worth revisiting as a rule rather than per-PR: it traces to `pipelex/kit/agent_rules/pytest_standards.md:80`, which we ship to users, and it has now been raised on three consecutive PRs. Captured with the measurements in `wip/pr-1139-review-notes.md`.

File: tests/e2e/pipelex/cli/test_validate_cmd.py [hunk 5]
```diff
@@ -1,11 +1,29 @@
+"""Smoke-test for the `validate --all` CLI path: a directory of bundles validates and dry-runs.
```
[the remark this answers is not in the document]
Comment (lchoquel) [reply tests/e2e/pipelex/cli/test_validate_cmd.py:1]: ➖ **False positive** — same population and same reading as the sibling thread on the parity module.

**The population.** 691 of 1130 test modules (61%) open with a module-level docstring; 9 of 13 in this file's own directory, `tests/e2e/pipelex/cli/`. The immediate neighbour `test_fix_bundle_cmd.py` opens with one *and* inlines a `_FIXABLE_SEQUENCE_MTHDS` bundle literal — both patterns already shipped side by side. Nothing enforces the rule: `pyproject.toml` ignores `D100`/`D101` and no rule forbids a module docstring.

**What the docstring is doing.** Not "this test tests X". It records that the directory is now assembled from corpus entries selected by `EntryValidity.VALID` rather than pointed at a path on disk, and why the old `Path(__file__).parents[4] / "pipelex"` expression became wrong the moment the corpus grew deliberately invalid entries. Deleting that invites the next reader to "simplify" the `copytree` loop straight back into the bug this PR just fixed. It also scopes the test against the corpus entry-validation gate, which is the stronger check — that boundary is worth stating where someone deciding what to add here will read it.

For what it is worth, the local norm in this feature area is the same: `tests/unit/pipelex/test_extras/test_mthds_corpus_exhaustivity.py` carried a module docstring naming its governing spec section before this PR touched it, and all four modules in that package do.

The one fair nit is that `test_validate_all` has no method docstring, so unlike the parity module the purpose sits only at module level — a real, if small, asymmetry with the rule's positive half. Not spending a change on it here, since it is cosmetic and this file would be starting to observe a rule 439 other modules do not.

Captured with the measurements, and with the two prior rounds of the same finding, in `wip/pr-1139-review-notes.md`.
